### ISSUE TITLE
need equivalent to invoice numbers for warehouse orders

### ISSUE THREAD

**Issue by babybunny** [babybunny/rebuildingtogethercaptain#297]
auto-generated number to identify orders. so it can appear in CDBG reporting.  

should appear in the reporting summary and also on top of the order reconciliation.

**Comment by babybunny** [comment]
need a special page for RTP warehouse orders. 

on order_fulfill page, if the order's vendor is RTP warehouse then have a special button that says "print invoice" and go to a similar page "Rebuilding Together Peninsula - Warehouse Invoice" and it should have a  
date
supplier (payable to:)
address
invoice number

**Comment by babybunny** [comment]
so, I understand also that the invoice numbers should be incrementing integers (which is standard for invoice numbers)

and maybe they are offset from some starting point like 10000, similar to my checkbook.

one great way to construct invoice IDs would be to use the beginning of the site ID (like 176001) and then add a 3 - or 4-digit serial number.

therefore we will need to store the invoice number somewhere. it could be simply a field on the Order model.  we can have the OrderSheet include a boolean that determines whether it can have an  internal invoice.  or a String which is more clear ("InvoicingMode" which is a choice).

on the order list page we will have a button to print the invoices for the orders that are associated with ordersheets that have an "internal" value for the invoice support.

**Comment by babybunny** [comment]
we can put the most recent invoice number on the site.

**Comment by babybunny** [comment]
Or we can just keep a global invoice number.  or one per program. 

I'm not convinced that a global is any worse than any of the other options.

If orders move between sites or programs it will be a mess to try to keep the invoice numbers in line with any site- or program-specific convention.  So I'm leaning towards just having a global.

**Comment by andy-d** [comment]
Yeah I like global, no reason they need encode useful information right?

**Comment by babybunny** [comment]
added a global InvoiceNumber object

dev~None> ndb_models.InvoiceNumber(id='global', next_invoice_number=10000).put()
Key('InvoiceNumber', 'global')
dev~None> ndb.Key(ndb_models.InvoiceNumber, 'global').get()
InvoiceNumber(key=Key('InvoiceNumber', 'global'), next_invoice_number=10000)

**Comment by babybunny** [comment]
ran that on dev server with 
$ python /Users/babybunny/google-cloud-sdk/platform/google_appengine/remote_api_shell.py -s localhost:8082

**Comment by babybunny** [comment]
so, now I'm learning about transactions.  
https://cloud.google.com/appengine/docs/standard/python/ndb/transactions

need to sync on some ancestor.   can use a cross-group (xg=True) transaction but it's not supported in the dev server, so can't test it.  

I discovered this when I wrote a test and it failed because it was xg and when I allowed the xg=True transaction it failed again because xg is not allowed.  So you must use xg but xg is not allowed.

anyway, now I have three choices.  

1) I can give up on the transaction (and hope that I don't get duplicate invoice numbers, which is incorrect).  
2) I can give up on testing this locally and just hope that I can demonstrate its correctness in prod.  Seeing that I don't have experience with this type of code this seems stupid.  
3) I can re-architect the thing to use a more practical transactional domain.  thinking that I can create an OrderInvoice object and make it a child of an InvoiceNumber.  Then I can synchronize on the InvoiceNumber which is what I want to do anyway. The Order can still store a copy of the invoice number for convenience. But it's also pretty easy to look up an OrderInvoice object by Order key and get the number from that. 

So, when user clicks a "Create Invoice" button, we'll assign an invoice number and take them to a printout page.  This might take a couple of seconds.

**Comment by babybunny** [comment]
Pushed a version of this to prod.  need to create the InvoiceNumber object now.

**Comment by babybunny** [comment]
ok, done.  I can see invoice number 10000!

https://rebuildingtogethercaptain.appspot.com/room/order_invoice/4754776076058624

The way this works is that there is a new field on the Order Form, called "supports_internal_invoice".
I set it for the Healthy Housing form.

When this is set then a link will appear on the reconcile page for orders on that form.  For example: https://rebuildingtogethercaptain.appspot.com/room/order_reconcile/6340104020819968

**Comment by MaryVeliz** [comment]
Good Morning,

I noticed that the healthy housing orders have an option to print an invoice. Can you also add the invoice # and option for all orders stemming from the RTP warehouse for billing purposes?
